packages/optimistic-oracle/src/proposer.js

```diff
@@ -157,80 +159,77 @@ class OptimisticOracleProposer {
       return;
     }
 
-    // Create the transaction.
+    // Get successful transaction receipt and return value or error.
     const proposal = this.optimisticOracleContract.methods.proposePrice(
       priceRequest.requester,
       this.utf8ToHex(priceRequest.identifier),
       priceRequest.timestamp,
       priceRequest.ancillaryData,
       proposalPrice
     );
-
-    // Simple version of inventory management: simulate the transaction and assume that if it fails,
-    // the caller didn't have enough collateral.
-    let proposalBond, gasEstimation;
-    try {
-      [proposalBond, gasEstimation] = await Promise.all([
-        proposal.call({ from: this.account }),
-        proposal.estimateGas({ from: this.account })
-      ]);
-    } catch (error) {
-      this.logger.error({
-        at: "OptimisticOracleProposer#sendProposals",
-        message: "Cannot propose price: not enough collateral (or large enough approval)✋",
-        proposer: this.account,
-        priceRequest,
-        error
-      });
-      return;
-    }
     const txnConfig = {
-      from: this.account,
-      gas: Math.min(Math.floor(gasEstimation * this.GAS_LIMIT_BUFFER), this.txnGasLimit),
       gasPrice: this.gasEstimator.getCurrentFastPrice()
     };
-
     this.logger.debug({
       at: "OptimisticOracleProposer#sendProposals",
       message: "Proposing new price",
       priceRequest,
-      proposalBond,
       proposalPrice,
       txnConfig
     });
+    const transactionResult = await runTransaction(
+      proposal,
+      this.account,
+      txnConfig,
+      this.GAS_LIMIT_BUFFER,
+      this.txnGasLimit
+    );
 
-    // Send the transaction or report failure.
-    let receipt;
-    try {
-      receipt = await proposal.send(txnConfig);
-    } catch (error) {
-      this.logger.error({
+    // Handle error:
+    if (transactionResult.error) {
+      if (transactionResult.error.type === "call") {
+        this.logger.error({
+          at: "OptimisticOracleProposer#sendProposals",
+          message: "Cannot propose price: not enough collateral (or large enough approval)✋",
+          priceRequest,
+          error: transactionResult.error
+        });
+        return;
+      } else {
+        this.logger.error({
+          at: "OptimisticOracleProposer#sendProposals",
+          message: "Failed to propose price🚨",
+          priceRequest,
+          error: transactionResult.error
+        });
+        return;
+      }
```

Comment:
I think you can shorten this because the log is exactly the same except for the message (you should be able to do the same elsewhere).
```suggestion
        const message = transactionResult.error.type === "call" ? "Cannot propose price: not enough collateral (or large enough approval)✋" : "Failed to propose price🚨";
        this.logger.error({
          at: "OptimisticOracleProposer#sendProposals",
          message,
          priceRequest,
          error: transactionResult.error
        });
        return;
```

Reply:
Great optimization